src/storage-types/src/read_holds.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fmt::Debug;
+
+use mz_repr::GlobalId;
+use timely::progress::{Antichain, ChangeBatch, Timestamp as TimelyTimestamp};
+use timely::PartialOrder;
+use tokio::sync::mpsc::UnboundedSender;
+
+/// Token that represents a hold on a collection. This prevents the since of the
+/// collection from progressing beyond the hold. In other words, it cannot
+/// become true that our hold is `less_than` the since.
+///
+/// This [ReadHold] is safe to drop. The installed read hold will be returned to
+/// the issuer behind the scenes.
+pub struct ReadHold<T: TimelyTimestamp> {
+    /// Identifies that collection that we have a hold on.
+    pub id: GlobalId,
+
+    /// The times at which we hold.
+    pub since: Antichain<T>,
+
+    /// For communicating changes to this read hold back to whoever issued it.
+    pub holds_tx: UnboundedSender<(GlobalId, ChangeBatch<T>)>,
```

Comment:
I think these fields shouldn't be `pub`?

Reply:
I had them pub because I was constructing these in tests. Later I added `new()` so can now remove them. 👌